### PR TITLE
An auto recommendation for inverted index

### PR DESCRIPTION
**Pull request by jasperjiaguo**
A solution to issue https://github.com/apache/incubator-pinot/issues/4461

### CODE REVIEW

File: pinot-tools/src/main/java/org/apache/pinot/tools/admin/PinotAdministrator.java [hunk 1]
```diff
@@ -80,7 +83,7 @@
...
...name = "DeleteCluster", impl = DeleteClusterCommand.class), @SubCommand(name = "ShowClusterInfo", impl = ShowClusterInfoCommand.class), @SubCommand(name = "AvroSchemaToPinotSchema", impl = AvroSchemaToPinotSchema.class), @SubCommand(name = "RebalanceTable", impl = RebalanceTableCommand.class), @SubCommand(name = "ChangeNumReplicas", impl = ChangeNumReplicasCommand.class), @SubCommand(name = "ValidateConfig", impl = ValidateConfigCommand.class), @SubCommand(name = "VerifySegmentState", impl = VerifySegmentState.class), @SubCommand(name = "ConvertPinotSegment", impl = PinotSegmentConvertCommand.class), @SubCommand(name = "MoveReplicaGroup", impl = MoveReplicaGroup.class), @SubCommand(name = "BackfillSegmentColumn", impl = BackfillDateTimeColumnCommand.class), @SubCommand(name = "VerifyClusterState", impl = VerifyClusterStateCommand.class), @SubCommand(name = "ApplyTableConfig", impl = ApplyTableConfigCommand.class), @SubCommand(name = "RealtimeProvisioningHelper", impl = RealtimeProvisioningHelperCommand.class), @SubCommand(name = "MergeSegments", impl = SegmentMergeCommand.class), @SubCommand(name = "CheckOfflineSegmentIntervals", impl = OfflineSegmentIntervalCheckerCommand.class)})
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/admin/PinotAdministrator.java]: Let's name the commands something more self explanatory. How about, `IndexTunerCommand` or `IndexRecommendorCommand`? `CollectSegmentMetadataForTuningCommand`? and `QueryLatencyQuantileReport`? What do you think?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/CollectMetaCommand.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
+package org.apache.pinot.tools.tuner;
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/CollectMetaCommand.java]: Add license header to all new files

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/CollectMetaCommand.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      tableNamesWithoutType.addAll(Arrays.asList(_tableNamesWithoutType.split(",")));
+    }
+
+    TunerDriver metaFetch = new TunerDriver().setThreadPoolSize(Runtime.getRuntime().availableProcessors() - 1)
+        .setStrategy(new AccumulateStats.Builder()
+            .setTableNamesWithoutType(tableNamesWithoutType)
+            .setOutputDir(_workDir)
+            .build())
+        .setQuerySrc(new CompressedFilePathIter.Builder()
+            .set_directory(_segmentsDir)
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/CollectMetaCommand.java]: method name should be camelCase, s/set_directory/setDirectory

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/TunerCommand.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+
+public class TunerCommand extends AbstractBaseCommand implements Command {
+  private static final long DEFAULT_NUM_ENTRIES_SCANNED_THRESHOLD = 0;
+  private static final long DEFAULT_NUM_QUERIES_TO_GIVE_RECOMMENDATION = 0;
+
+  private static final String INVERTED_INDEX = "inverted";
+  private static final String SORTED_INDEX = "sorted";
+
+  @Option(name = "-metaData", required = true, metaVar = "<String>", usage = "Path to packed metadata file (json).")
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/TunerCommand.java]: Add to this description, that the CollectMetaCommand can be used to create this

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/TunerCommand.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  @Option(name = "-brokerLog", required = true, metaVar = "<String>", usage = "Path to broker log file.")
+  private String _brokerLog;
+
+  @Option(name = "-strategy", required = true, metaVar = "<inverted/sorted>", usage = "Select execution strategy.")
+  private String _strategy;
+
+  @Option(name = "-entriesScannedThreshold", required = false, metaVar = "<long>", usage = "Log lines with numEntriesScannedInFilter below this threshold will be excluded.")
+  private long _numEntriesScannedThreshold = DEFAULT_NUM_ENTRIES_SCANNED_THRESHOLD;
+
+  @Option(name = "-queriesToReport", required = false, metaVar = "<long>", usage = "Log lines with numEntriesScannedInFilter below this threshold will be excluded.")
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/TunerCommand.java]: This description looks like it's been copied over from entriedScannedThreshold?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/TunerCommand.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      parserBased.execute();
+    } else {
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  public String description() {
+    return "Give optimization boundary analysis and indexing recommendation to specific tables, based on packed segment metadata (containing the weighted sum of cardinality, number of documents, number of entries, etc.) and broker logs (containing time of execution, numEntriesScannedInFilter, numEntriesScannedPostFilter, query text).";
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/TunerCommand.java]: Not sure what "give optimization boundary analysis" means.. can you make this bit simpler?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/TunerCommand.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    HashSet<String> tableNamesWithoutType = new HashSet<>();
+    if (_tableNamesWithoutType != null && !_tableNamesWithoutType.trim().equals("")) {
+      tableNamesWithoutType.addAll(Arrays.asList(_tableNamesWithoutType.split(",")));
+    }
+
+    if (_strategy.equals(INVERTED_INDEX)) {
+      TunerDriver parserBased = new TunerDriver().setThreadPoolSize(Runtime.getRuntime().availableProcessors() - 1)
+          .setStrategy(new ParserBasedImpl.Builder()
+              .setTableNamesWithoutType(tableNamesWithoutType)
+              .setNumProcessedThreshold(_numQueriesToGiveRecommendation)
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/TunerCommand.java]: 3 different names to refer to the same thing -> _numQueriesToGiveRecommendation, numProcessedThreshold, queriesToReport. Let's make it uniform, I prefer numQueriesThreshold

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/Strategy.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+   * @param queryStats the stats extracted and parsed from QuerySrc
+   * @return
+   */
+  boolean filter(AbstractQueryStats queryStats);
+
+  /**
+   * Accumulate the parsed queryStats to corresponding entry in MapperOut, see FrequencyImpl for ex
+   * @param queryStats input, the stats extracted and parsed from QuerySrc
+   * @param metaManager input, the metaManager where cardinality info can be get from
+   * @param AccumulatorOut output, map of /tableMame: String/columnName: String/AbstractMergerObj
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/Strategy.java]: AccumulatorOut -> accumulatorOut

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/driver/TunerDriver.java [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+   * Set the strategy for the recommendation, e.g. FrequencyImpl, OLSAnalysisImpl, ParserBasedImpl
+   * @param strategy E.g. ParserBasedImpl
+   * @return this
+   */
+  public TunerDriver setStrategy(Strategy strategy) {
+    _strategy = strategy;
+    return this;
+  }
+
+  private Map<Long, Map<String, Map<String, AbstractAccumulator>>> _threadAccumulator = null;
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/driver/TunerDriver.java]: multiple threads are accessing the same Map. Don't you need a ConcurrentHashMap?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+  public void report(Map<String, Map<String, AbstractAccumulator>> tableResults) {
+    tableResults.forEach((table, map) -> {
+      reportTable(table, map);
+    });
+  }
+
+  public void reportTable(String tableNameWithoutType, Map<String, AbstractAccumulator> columnStats) {
+    long totalCount = columnStats.remove(NUM_QUERIES_COUNT).getCount();
+    if (totalCount < _numQueriesThreshold) {
+      return;
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java]: return without any message?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+    while (matcher.find()) {
+      if (matcher.group(1) != null) {
+        counted.add(matcher.group(1));
+      } else if (matcher.group(4) != null) {
+        counted.add(matcher.group(4));
+      } else {
+      }
+    }
+
+    counted.stream().filter(colName -> metaManager.getColumnSelectivity(tableNameWithoutType, colName)
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java]: even the simple frequency based requires the metadata manager? no option to skip it?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+        .compareTo(new BigFraction(_cardinalityThreshold)) > 0).forEach(colName -> {
+      accumulatorOut.putIfAbsent(tableNameWithoutType, new HashMap<>());
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java]: didn't we already putIfAbsent(tableNameWithoutType on line 143?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+      accumulatorOut.get(tableNameWithoutType).putIfAbsent(colName, new FrequencyAccumulator());
+      ((FrequencyAccumulator) accumulatorOut.get(tableNameWithoutType).get(colName)).merge(1);
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java]: merge(1)?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyAccumulator.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.tools.tuner.strategy;
+
+/**
+ *
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyAccumulator.java]: empty?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyAccumulator.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+ */
+public class FrequencyAccumulator extends AbstractAccumulator {
+
+  public FrequencyAccumulator() {
+    _pureScore = 0;
+  }
+
+  private long _pureScore;
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyAccumulator.java]: this is simply a frequency or a count right? Why is it called pureScore? Add a comment if this needs to be called pureScore.

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/AbstractAccumulator.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    return _count;
+  }
+
+  private long _count = 0;
+
+  /**
+   * Increase default counter by one
+   */
+  public void increaseCount() {
+    this._count += 1;
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/AbstractAccumulator.java]: no need of `this.`

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/AbstractAccumulator.java [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+package org.apache.pinot.tools.tuner.strategy;
+
+/**
+ * Accumulator for column stats
+ */
+public abstract class AbstractAccumulator {
+  public abstract String toString();
+
+  /**
+   * Get the default counter for BasicMergerObjs merged to this AbstractMergerObj
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/AbstractAccumulator.java]: What is BasicMergerObj and AbstractMergerObj? Maybe reword this to "get count of number of AbstractAccumulator objects accumulated.

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/query/src/LogQuerySrcImpl.java [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+
+  @Override
+  public AbstractQueryStats next()
+      throws NoSuchElementException {
+    if (_stringBufferNext == null) {
+      throw new NoSuchElementException();
+    }
+    StringBuilder stringBuffer = new StringBuilder(_stringBufferNext);
+    try {
+      while ((_stringBufferNext = _bufferedReader.readLine()) != null && !_validLinePrefixPattern
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/query/src/LogQuerySrcImpl.java]: add a comment about why we're still reading more even if we find a line, it reads a little confusing

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+      _numQueriesThreshold = val;
+      return this;
+    }
+  }
+
+  @Override
+  public boolean filter(AbstractQueryStats queryStats) {
+    IndexSuggestQueryStatsImpl indexSuggestQueryStatsImpl = (IndexSuggestQueryStatsImpl) queryStats;
+    long numEntriesScannedInFilter = Long.parseLong(indexSuggestQueryStatsImpl.getNumEntriesScannedInFilter());
+    return (_tableNamesWithoutType == null || _tableNamesWithoutType.isEmpty() || _tableNamesWithoutType
```
[the remark this answers is not in the document]
Comment (jasperjiaguo) [reply pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java]: We will not skip the numEntriesScannedThreshold check because it's (_tableNamesWithoutType == null || ... || ...) **&&** (numEntriesScannedInFilter > _numEntriesScannedThreshold)

[the remark this answers is not in the document]
Comment (jasperjiaguo) [reply pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java]: > why is filter method being made to check for _tableNamesWithoutType == null || _tableNamesWithoutType.isEmpty? This is going to be repeated simply for every query. Can it be done before the 4 stages begin?

Good point, I will make it in the builder of strategy.

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java (hunk 11 above)
Comment (jasperjiaguo) [reply pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java]: This is slightly more intelligent than very simple frequency, we can set a cardinality threshold for the simple counter, which is more reasonable

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java (hunk 12 above)
Comment (jasperjiaguo) [reply pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/FrequencyImpl.java]: I have put all these repetitive putIfAbsent stuff a function called putAccumulatorToMapIfAbsent in AbstractAccumulator class

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/OLSAccumulator.java [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+package org.apache.pinot.tools.tuner.strategy;
+
+import io.vavr.Tuple2;
+import java.util.ArrayList;
+import java.util.HashMap;
+
+
+/**
+ * An implementation of {@link AbstractAccumulator}, to implement minimum pulling for {@link OLSAnalysisImpl}
+ */
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/OLSAccumulator.java]: What does OLS mean?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/OLSAnalysisImpl.java [hunk 21]
```diff
@@ -0,0 +1,348 @@
...
+    String numEntriesScannedInFilter = indexSuggestQueryStatsImpl.getNumEntriesScannedInFilter();
+    String numEntriesScannedPostFilter = indexSuggestQueryStatsImpl.getNumEntriesScannedPostFilter();
+    String query = indexSuggestQueryStatsImpl.getQuery();
+    LOGGER.debug("Accumulator: scoring query {}", query);
+
+    accumulatorOut.putIfAbsent(tableNameWithoutType, new HashMap<>());
+    accumulatorOut.get(tableNameWithoutType).putIfAbsent(NUM_QUERIES_COUNT, new OLSAccumulator());
+    accumulatorOut.get(tableNameWithoutType).get(NUM_QUERIES_COUNT).increaseCount();
+
+    accumulatorOut.get(tableNameWithoutType).putIfAbsent("*", new OLSAccumulator());
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/OLSAnalysisImpl.java]: Make this a proper key name instead of using "*"

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/OLSAnalysisImpl.java [hunk 22]
```diff
@@ -0,0 +1,348 @@
...
+  }
+
+  public void reportTable(String tableNameWithoutType, Map<String, AbstractAccumulator> columnStats) {
+    String reportOut = "\n**********************Report For Table: " + tableNameWithoutType + "**********************\n";
+
+    long totalCount = columnStats.remove(NUM_QUERIES_COUNT).getCount();
+    reportOut += MessageFormat.format("\nTotal lines accumulated: {0}\n\n", totalCount);
+
+    if (!columnStats.containsKey("*")) {
+      return;
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/OLSAnalysisImpl.java]: add a message before return so that we know why we dont have a report

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/OLSAnalysisImpl.java [hunk 23]
```diff
@@ -0,0 +1,348 @@
...
+    } catch (MathIllegalArgumentException | MathArithmeticException e) {
+      reportOut += "\nunable to predict the optimization boundary of this table!";
+    }
+    LOGGER.info(reportOut);
+  }
+
+  /**
+   * Parse and count the total bitmaps used in a query
+   */
+  class DimensionScoring {
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/OLSAnalysisImpl.java]: why is DimensionScoring here? I don't see it being used in this class anywhere

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/query/src/parser/BrokerLogParserImpl.java [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+ * Implementation of {@link QueryParser}, to parse complete lines of broker log text
+ */
+public class BrokerLogParserImpl implements QueryParser {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BrokerLogParserImpl.class);
+
+  //Regex to parse Broker Log
+  private static final String BROKER_LOG_REGEX =
+      "^.*?table:(.+?)(?:_OFFLINE|_REALTIME|_HYBRID|)?, timeMs:(\\d+), docs:(\\d+)/(\\d+).*, entries:(\\d+)/(\\d+),.*, query:(.+)$";
+
+  private static final Pattern _compiledPattern = Pattern.compile(BROKER_LOG_REGEX);
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/query/src/parser/BrokerLogParserImpl.java]: generally, static final fields are written in caps COMPILED_PATTERN

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/query/src/parser/BrokerLogParserImpl.java [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+
+  private enum GROUP_NAMES {
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/query/src/parser/BrokerLogParserImpl.java]: enum name should be camel case: GroupNames

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/CompressedFilePathIter.java [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+  private CompressedFilePathIter openDirectory() {
+    File dir = new File(_directory);
+    if (!dir.exists() || dir.isFile()) {
+      LOGGER.error("Wrong input directory!");
+      System.exit(1);
+    }
+
+    ArrayList<Tuple2<String, File>> validTableNameWithoutTypeSegmentFile = new ArrayList<>();
+
+    Arrays.stream(Objects.requireNonNull(dir.listFiles()))
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/CompressedFilePathIter.java]: This line reads too long. I suggest break it up into 2 sweeps.
And the validations for "." and "DeletedSegments" are too specific.

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/CompressedFilePathIter.java [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+
+
+/**
+ * The iterator over all tarred segments in _directory/tableName/
+ */
+public class CompressedFilePathIter implements QuerySrc {
+  private static final Logger LOGGER = LoggerFactory.getLogger(CompressedFilePathIter.class);
+  Iterator<Tuple2<String, File>> _iterator;
+
+  private String _directory;
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/CompressedFilePathIter.java]: rename to _segmentsDir or _segmentTarDir

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/CompressedFilePathIter.java [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The iterator over all tarred segments in _directory/tableName/
+ */
+public class CompressedFilePathIter implements QuerySrc {
+  private static final Logger LOGGER = LoggerFactory.getLogger(CompressedFilePathIter.class);
+  Iterator<Tuple2<String, File>> _iterator;
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/CompressedFilePathIter.java]: private?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/CompressedFilePathIter.java [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+  private String _directory;
+
+  private CompressedFilePathIter(Builder builder) {
+    _directory = builder._directory;
+  }
+
+  private CompressedFilePathIter openDirectory() {
+    File dir = new File(_directory);
+    if (!dir.exists() || dir.isFile()) {
+      LOGGER.error("Wrong input directory!");
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/CompressedFilePathIter.java]: can we bubble this up to Driver/Command class instead of System.exit (for all similar usages), and handle the exceptions in that entry level class?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/ColStatsAccumulatorObj.java [hunk 30]
```diff
@@ -0,0 +1,142 @@
...
+    return this;
+  }
+
+  public ColStatsAccumulatorObj addIsSorted(String isSorted) {
+    _isSorted = isSorted;
+    return this;
+  }
+
+  public void merge() {
+    accumulatedStats.put(MetaManager.WEIGHTED_SUM_CARDINALITY,
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/ColStatsAccumulatorObj.java]: I suggest introducing a method
```
private void addToAccumulatedStats(String key,  BigInteger value) {
  if (!_accumulatedStats.containsKey(key)) {
    _accumulatedStats.put(key, BigInteger.ONE);
  }
  _accumulatedStats.get(key).add(value);
}
```
and then simply call
```
BigInteger weightedSumCardinality = new BigInteger(_cardinality).multiply(new BigInteger(_totalDocs));
addToAccumulatedStats(MetaManager.WEIGHTED_SUM_CARDINALITY, weightedSumCardinality);
```
and follow the same in the other `merge(ColStatsAccumulator obj)` method.

```

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/ColStatsAccumulatorObj.java [hunk 31]
```diff
@@ -0,0 +1,142 @@
...
+import org.apache.pinot.tools.tuner.meta.manager.MetaManager;
+import org.apache.pinot.tools.tuner.strategy.AbstractAccumulator;
+
+
+/**
+ * Accumulator for relevant fields in metadata.properties and index_map
+ */
+public class ColStatsAccumulatorObj extends AbstractAccumulator {
+
+  Map<String, BigInteger> accumulatedStats = new HashMap<>();
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/ColStatsAccumulatorObj.java]: _accumulatedStats
_segmentStats

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/ColStatsAccumulatorObj.java [hunk 32]
```diff
@@ -0,0 +1,142 @@
...
+    _totalNumberOfEntries = totalNumberOfEntries;
+    return this;
+  }
+
+  public ColStatsAccumulatorObj addIsSorted(String isSorted) {
+    _isSorted = isSorted;
+    return this;
+  }
+
+  public void merge() {
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/ColStatsAccumulatorObj.java]: why do we have a `merge()` method which is not taking any inputs to merge, but instead operating on the fields of the object? Instead, can we pass all those numbers as arguments to this method, so we know exactly what we're merging?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/ColStatsAccumulatorObj.java [hunk 33]
```diff
@@ -0,0 +1,142 @@
...
+
+  public Map<String, BigInteger> getAccumulatedStats() {
+    return accumulatedStats;
+  }
+
+  public Map<String, Map<String, String>> getSegmentStats() {
+    return segmentStats;
+  }
+
+  public ColStatsAccumulatorObj addInvertedIndexSize(String invertedIndexSize) {
```
Comment (npawar) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/ColStatsAccumulatorObj.java]: these methods are not really adding the numbers, but instead just setting the values to the fields. Why do we need these? these should be params to the merge method right?

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/ColStatsAccumulatorObj.java (hunk 32 above)
Comment (jasperjiaguo) [reply pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/ColStatsAccumulatorObj.java]: I wrote it this way so that when merging the code is like following, which I think is more readable than a function with a long parameter list
          ```
.addInvertedIndexSize(invertedIndexSize)
          .addIsSorted(isSorted)
          .addSegmentName(pathWrapper.getFile().getName())
          .addTotalDocs(totalDocs)
          .addTotalNumberOfEntries(totalNumberOfEntries)
          .merge();
```

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/OLSAccumulator.java (hunk 20 above)
Comment (jasperjiaguo) [reply pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/OLSAccumulator.java]: OLS is Ordinary Least Squares regression, I will change the naming

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/query/src/LogQuerySrcImpl.java [hunk 34]
```diff
@@ -0,0 +1,155 @@
...
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Generic class to iterate over lines in file
+ */
+public class LogQuerySrcImpl implements QuerySrc {
+  private static final Logger LOGGER = LoggerFactory.getLogger(LogQuerySrcImpl.class);
+
+  public static final String REGEX_VALID_LINE_STANDALONE = "^(Processed requestId|RequestId).*$";
```
[the remark this answers is not in the document]
Comment (jasperjiaguo) [reply pinot-tools/src/main/java/org/apache/pinot/tools/tuner/query/src/LogQuerySrcImpl.java]: It's for both broker and server log

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/query/src/parser/ServerLogParserImpl.java [hunk 35]
```diff
@@ -0,0 +1,67 @@
...
+import org.apache.pinot.tools.tuner.query.src.stats.wrapper.AbstractQueryStats;
+import org.apache.pinot.tools.tuner.query.src.stats.wrapper.IndexSuggestQueryStatsImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Implementation of {@link QueryParser}, to parse complete lines of server log text
+ */
+public class ServerLogParserImpl implements QueryParser {
```
[the remark this answers is not in the document]
Comment (jasperjiaguo) [reply pinot-tools/src/main/java/org/apache/pinot/tools/tuner/query/src/parser/ServerLogParserImpl.java]: Not for now, but could be very useful to study the overhead of or-ing a very long list of inverted indices

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/OLSAnalysisImpl.java (hunk 23 above)
Comment (jasperjiaguo) [reply pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/OLSAnalysisImpl.java]: Not for now but will be useful for future study of overhead of or-ing a very long list of inverted indices

Comment (jasperjiaguo) [reply pinot-tools/src/main/java/org/apache/pinot/tools/tuner/strategy/OLSAnalysisImpl.java]: Not for now, but could be very useful to study the overhead of or-ing a very long list of inverted indices

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/ColStatsAccumulatorObj.java (hunk 30 above)
Comment (jasperjiaguo) [reply pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/collector/ColStatsAccumulatorObj.java]: Yes I will do it later when we have auto testing for this part. I fear it will break the data

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/JsonFileMetaManagerImpl.java [hunk 36]
```diff
@@ -0,0 +1,245 @@
...
+    }
+    return false;
+  }
+
+  public BigFraction getAverageNumEntriesPerDoc(String tableNameWithoutType, String columnName) {
+    try {
+      BigInteger entriesNumerator = new BigInteger(getColField(tableNameWithoutType, columnName, SUM_TOTAL_ENTRIES));
+      BigInteger entriesDenominator = new BigInteger(getColField(tableNameWithoutType, columnName, SUM_DOCS));
+      if (entriesNumerator.compareTo(entriesDenominator) < 0 || entriesDenominator.equals(BigInteger.ZERO)) {
+        throw new Exception();
```
Comment (sunithabeeram) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/JsonFileMetaManagerImpl.java]: Would be good to throw a meaningful exception - or at the very least include a message in the exception.

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/JsonFileMetaManagerImpl.java [hunk 37]
```diff
@@ -0,0 +1,245 @@
...
+      }
+      return new BigFraction(entriesNumerator, entriesDenominator);
+    } catch (Exception e) {
+      LOGGER.trace("Set average entries to 1 {} {}", tableNameWithoutType, columnName);
+      return BigFraction.ONE;
+    }
+  }
+
+  public BigFraction getColumnSelectivity(String tableNameWithoutType, String columnName) {
```
Comment (sunithabeeram) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/meta/manager/JsonFileMetaManagerImpl.java]: I think this is an important method - adding some comments/rationale around how sorted/cardinality signals are used would be helpful.

File: pinot-tools/src/main/java/org/apache/pinot/tools/tuner/query/src/parser/BrokerLogParserImpl.java [hunk 38]
```diff
@@ -0,0 +1,76 @@
...
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Implementation of {@link QueryParser}, to parse complete lines of broker log text
+ */
+public class BrokerLogParserImpl implements QueryParser {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BrokerLogParserImpl.class);
+
+  //Regex to parse Broker Log
```
Comment (mcvsubbu) [inline pinot-tools/src/main/java/org/apache/pinot/tools/tuner/query/src/parser/BrokerLogParserImpl.java]: A change to the log line will break the tool?

Perhaps the log line parers should be in the respective packages (broker/server)?

Bonus points if we can add a unit test that attempts to parse the log line generated by the current code, so we know we have broken it.
